File: src/lexer/lexer.rs
use anyhow::Result;

#[derive(Debug, PartialEq)]
pub enum Token {
    Ident(String),
    Int(String),

    Illegal,
    Eof,
    Equal,
    Plus,
    Comma,
    Semicolon,
    Lparen,
    Rparen,
    LSquirly,
    RSquirly,
    Function,
    Let,
}

pub struct Lexer {
    position: usize,
    read_position: usize,
    ch: u8,
    input: Vec<u8>,
}

impl Lexer {
    fn new(input: String) -> Lexer {
        let mut lex = Lexer {
            position: 0,
            read_position: 0,
            ch: 0,
            input: input.into_bytes(),
        };
        lex.read_char();

        return lex;
    }

    pub fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace();

        let tok = match self.ch {
            b'{' => Token::LSquirly,
            b'}' => Token::RSquirly,
            b'(' => Token::Lparen,
            b')' => Token::Rparen,
            b',' => Token::Comma,
            b';' => Token::Semicolon,
            b'+' => Token::Plus,
            b'=' => Token::Equal,
            b'a'..=b'z' | b'A'..=b'Z' | b'_' => {
                let ident = self.read_ident();
                return Ok(match ident.as_str() {
                    "fn" => Token::Function,
                    "let" => Token::Let,
                    _ => Token::Ident(ident),
                });
            },
            b'0'..=b'9' => return Ok(Token::Int(self.read_int())),
            0 => Token::Eof,
            _ => todo!("we need to implement this....")
        };

        self.read_char();
        return Ok(tok);
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0;
        } else {
            self.ch = self.input[self.read_position];
        }

        self.position = self.read_position;
        self.read_position += 1;
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }

    fn read_ident(&mut self) -> String {
        let pos = self.position;
        while self.ch.is_ascii_alphabetic() || self.ch == b'_' {
            self.read_char();
        }

        return String::from_utf8_lossy(&self.input[pos..self.position]).to_string();
    }

    fn read_int(&mut self) -> String {
        let pos = self.position;
        while self.ch.is_ascii_digit() {
            self.read_char();
        }

        return String::from_utf8_lossy(&self.input[pos..self.position]).to_string();
    }
}

#[cfg(test)]
mod test {
    use anyhow::Result;

    use super::{Token, Lexer};

    #[test]
    fn get_next_token() -> Result<()> {
        let input = "=+(){},;";
        let mut lexer = Lexer::new(input.into());

        let tokens = vec![
            Token::Equal,
            Token::Plus,
            Token::Lparen,
            Token::Rparen,
            Token::LSquirly,
            Token::RSquirly,
            Token::Comma,
            Token::Semicolon,
        ];

        for token in tokens {
            let next_token = lexer.next_token()?;
            println!("expected: {:?}, received {:?}", token, next_token);
            assert_eq!(token, next_token);
        }


        return Ok(());
    }

    #[test]
    fn get_next_complete() -> Result<()> {
        let input = r#"let five = 5;
            let ten = 10;
            let add = fn(x, y) {
                x + y;
            };
            let result = add(five, ten);"# ;

        let mut lex = Lexer::new(input.into());

        let tokens = vec![
            Token::Let,
            Token::Ident(String::from("five")),
            Token::Equal,
            Token::Int(String::from("5")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("ten")),
            Token::Equal,
            Token::Int(String::from("10")),
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("add")),
            Token::Equal,
            Token::Function,
            Token::Lparen,
            Token::Ident(String::from("x")),
            Token::Comma,
            Token::Ident(String::from("y")),
            Token::Rparen,
            Token::LSquirly,
            Token::Ident(String::from("x")),
            Token::Plus,
            Token::Ident(String::from("y")),
            Token::Semicolon,
            Token::RSquirly,
            Token::Semicolon,
            Token::Let,
            Token::Ident(String::from("result")),
            Token::Equal,
            Token::Ident(String::from("add")),
            Token::Lparen,
            Token::Ident(String::from("five")),
            Token::Comma,
            Token::Ident(String::from("ten")),
            Token::Rparen,
            Token::Semicolon,
            Token::Eof,
        ];

        for token in tokens {
            let next_token = lex.next_token()?;
            println!("expected: {:?}, received {:?}", token, next_token);
            assert_eq!(token, next_token);
        }

        return Ok(());
    }
}

File: lua/lexer/lexer.lua
#!/usr/bin/env luajit

local M = {}

M.token_t = {
  ASSIGN = 0,
  COMMA = 1,
  EOF = 2,
  FUNCTION = 3,
  IDENTIFIER = 4,
  ILLEGAL = 5,
  INTEGER = 6,
  LET = 7,
  LPAREN = 8,
  LSQUIRLY = 9,
  PLUS = 10,
  RPAREN = 11,
  RSQUIRLY = 12,
  SEMICOLON = 13,
}

M.lexer = function(str)
  if str == nil or str == "" then
    error("Empty Input Not Permitted", 1)
  end
  M.input = str
  M.pos = 1
  M.curr_byte = str:byte(1)
end

local read_next_byte = function()
  if M.pos >= #M.input then
    M.curr_byte = 0
    return
  end

  M.pos = M.pos + 1
  M.curr_byte = M.input:byte(M.pos)
end

local is_letter = function()
  return (97 <= M.curr_byte and M.curr_byte <= 122)
    or (65 <= M.curr_byte and M.curr_byte <= 90)
    or M.curr_byte == 95
end

local is_digit = function()
  return (48 <= M.curr_byte and M.curr_byte <= 57)
end

local get_number = function()
  local start = M.pos
  repeat
    read_next_byte()
  until not is_digit()
  return M.input:sub(start, M.pos - 1)
end

local get_identifier = function()
  local start = M.pos
  repeat
    read_next_byte()
  until not is_letter()
  return M.input:sub(start, M.pos - 1)
end

local skip_whitespace = function()
  while M.curr_byte == 10 or M.curr_byte == 9 or M.curr_byte == 13 or M.curr_byte == 32 do
    read_next_byte()
  end
end

local literal_to_token = {
  ["fn"] = M.token_t.FUNCTION,
  ["let"] = M.token_t.LET,
}

M.next_token = function()
  skip_whitespace()
  local token = { type = M.token_t.ILLEGAL, literal = string.char(M.curr_byte) }

  if token.literal == "=" then
    token.type = M.token_t.ASSIGN
  elseif token.literal == "+" then
    token.type = M.token_t.PLUS
  elseif token.literal == "\0" then
    token.type = M.token_t.EOF
    token.literal = "eof"
  elseif token.literal == ";" then
    token.type = M.token_t.SEMICOLON
  elseif token.literal == "," then
    token.type = M.token_t.COMMA
  elseif token.literal == "(" then
    token.type = M.token_t.LPAREN
  elseif token.literal == ")" then
    token.type = M.token_t.RPAREN
  elseif token.literal == "{" then
    token.type = M.token_t.LSQUIRLY
  elseif token.literal == "}" then
    token.type = M.token_t.RSQUIRLY
  end

  if is_digit() then
    return { type = M.token_t.INTEGER, literal = get_number() }
  end

  if is_letter() then
    local ident = get_identifier()
    local tok = literal_to_token[ident]
    if tok then
      return { type = tok, literal = ident }
    end
    return { type = M.token_t.IDENTIFIER, literal = ident }
  end

  read_next_byte()
  return token
end

return M
